src/coreclr/jit/compiler.h

```diff
@@ -9191,39 +9191,61 @@ XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX
 // Maximum number of locals before turning off the inlining
 #define MAX_LV_NUM_COUNT_FOR_INLINING 512
 
+    private:
         bool compMinOpts;
         bool compMinOptsIsSet;
+        bool compDbgCode;             // Generate debugger-friendly code?
+        bool compOptimizationEnabled; // A cached, composite value
+
 #ifdef DEBUG
         mutable bool compMinOptsIsUsed;
+#endif // !DEBUG
+
+        void SetOptimizationEnabled()
+        {
+            // We want to be careful with the `compMinOptsIsSet` and `compMinOptsIsUsed` values.
+            // Caching the value shouldn't touch these. However, using this value should have the
+            // same asserts as using `MinOpts()`.
+            compOptimizationEnabled = !compMinOpts && !compDbgCode;
+        }
+
+    public:
+        void InitializeMinOpts()
+        {
+            INDEBUG(compMinOptsIsUsed = false);
+            compMinOptsIsSet = false;
+        }
 
+#ifdef DEBUG
         bool MinOpts() const
         {
             assert(compMinOptsIsSet);
             compMinOptsIsUsed = true;
```

Comment:
Just wondering, what's the general point of this and why do we need to track it as part of the check whether min opts is enabled

Reply:
Agree. Shouldn't `compMinOptsIsSet` sufficient to make sure that we are not calling `MinOpts()` before it is set?

Reply:
Once we read the value, we should never try to set it afterwards -- someone has taken a dependency on the value. Like the PhasedVar stuff.

Reply:
Couldn't we just have a check in `SetMinOpts` to assert we aren't past `import` or anything?

We should really only be setting it when the compiler is created or in the case we fault and fallback to minopts, right?

Once we've hit import, we're basically guaranteed to have some path that checks it.

Reply:
Perhaps, but the effect would not be as obvious or direct, and of course, `compMinOptsIsUsed` is DEBUG-only code.